prov/efa/src/rxr/rxr_read.c

```diff
@@ -79,15 +79,67 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
 	return -1;
 }
 
+/*
+ * rxr_read_mr_reg register the memory of local buffer if application did not
+ * provide descriptor.
+ * It is called by rxr_read_post().
+ * On success, it return 0.
+ * If memory registration failed with -FI_ENOMEM, it will return -FI_EAGAIN.
+ * If memory registration failed with other error, it will return the error code.
+ */
+ssize_t rxr_read_mr_reg(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
+{
+	size_t i;
+	int err;
+
+	for (i = 0; i < read_entry->iov_count; ++i) {
+		if (read_entry->mr_desc[i] || read_entry->mr[i]) {
+			continue;
+		}
+
+		err = fi_mr_reg(rxr_ep_domain(ep)->rdm_domain,
+				read_entry->iov[i].iov_base, read_entry->iov[i].iov_len,
+				FI_RECV, 0, 0, 0, &read_entry->mr[i], NULL);
+
+		if (err) {
+			/* If registration failed with -FI_ENOMEM, we return -FI_EAGAIN.
+			 * The progress engine will release registrations and call this
+			 * function again.
+			 *
+			 * All registration opened here will be closed during release of
```

Comment:
Is this correct? Looks like it leaves the registrations alone and queues the message.